marshal/consumer.go

```diff
@@ -375,6 +394,35 @@ func (c *Consumer) claimTopics() {
 		}
 	}
 
+	// let's compare the new topic claims vs old ones. Upon change, we should trigger an update
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	changed := false
+	// check for missing topic claims
+	for topic := range c.claimedTopics {
+		if !latestClaims[topic] {
+			changed = true
+			break
+		}
+	}
+	// check for new topic claims
+	if !changed && len(latestClaims) != len(c.claimedTopics) {
+		changed = true
+	}
+
+	if changed {
+		c.lock.RUnlock()
```

Comment:
My main point is around callstacks. If you have a few places doing go foo() where foo is doing some locking operations, then you have no idea who did the operation because well...it's happening asynchronously on a go routine. It makes debugging code really hard. Whereas if you have it synchronously, then there is a very clear callstack you can debug.
Regarding acquiring the Lock, this is for performance reasons and also most existing code is already acquiring the RLock.